Numerics: BigInt Header

1010jms opened the pull request:
Created a header file for the BigInteger class that contains the definitions for the constructors as well as the getter, the conversion functions to Rational and Real, the basic arithmetic operators, and the comparison operators.

Created the initial cpp file that contains comments explaining which algorithms need to be implemented for each function (as well as a link to each algorithm) as well as which functions need to be tested and which functions still need to be figured out.

Created unit test file that successfully tests the BigInteger constructors.

btwooton commented on src/shaka_scheme/system/base/BigInteger.cpp:
```diff
@@ -0,0 +1,147 @@
+#include "shaka_scheme/system/base/Real.hpp"
+
+namespace shaka {
+    BigInteger::operator Rational() {
```
@1010jsm One potential issue with a direct conversion from a `BigInteger` to a `Rational` is that class `Rational` is implemented using `int` to encode the numerator and denominator. Since a `BigInteger` is conceptually converted to a `Rational` by giving its value as the numerator, and 1 as the denominator, this encoding will break down for a direct conversion from `BigInteger` to `Rational`, since the C++ `int` type wont be able to store the `BigInteger` value. Your heart is certainly in the right place, but this is an implementation detail that you will need to consider later. One, potentially naive, option to consider is to have a `BigRational` class that you can seamlessly convert to and from. The numerator and denominator of a `BigRational` would themselves be `BigInteger`s. Just wanted to chime in with that, but I like what I see so far as far as scaffolding goes.

1010jms commented:
Update: 
- Moved all of the function bodies in the header file to the cpp file, which is currently under construction
- Added a default constructor
- Added more comments for the constructors, arithmetic operators, and comparison operators
- Removed the cpp and unit-test file from this pull request. Will create dedicated pull requests for each file in the future.
